cmplxs/parse_test.go

```diff
@@ -0,0 +1,79 @@
+// Copyright ©2020 The Gonum Authors. All rights reserved.
+// Use of this code is governed by a BSD-style
+// license that can be found in the LICENSE file
+
+package cmplxs
+
+import (
+	"math"
+	"math/cmplx"
+	"testing"
+)
+
+var parseTests = []struct {
+	s       string
+	want    complex128
+	wantErr error
+}{
+	// Simple error states:
+	{s: "", wantErr: parseError{state: -1}},
+	{s: "()", wantErr: parseError{string: "()", state: -1}},
+	{s: "(1", wantErr: parseError{string: "(1", state: -1}},
+	{s: "1)", wantErr: parseError{string: "1)", state: -1}},
+
+	// Ambiguous parse error states:
+	{s: "1+2i+3i", wantErr: parseError{string: "1+2i+3i", state: -1}},
+	{s: "1e-4i+", wantErr: parseError{string: "1e-4i+", state: -1}},
+	{s: "1e-4i-", wantErr: parseError{string: "1e-4i-", state: -1}},
+
+	// Valid input:
+	{s: "1+4i", want: 1 + 4i},
+	{s: "4i+1", want: 1 + 4i},
+	{s: "+1+4i", want: 1 + 4i},
+	{s: "+4i+1", want: 1 + 4i},
+	{s: "1e-4+1i", want: 1e-4 + 1i},
+	{s: "1e-4+i", want: 1e-4 + 1i},
+	{s: "1e-4-i", want: 1e-4 - 1i},
+	{s: "1e-4i-1", want: -1 + 1e-4i},
+	{s: "1e-4i+1", want: 1 + 1e-4i},
+	{s: "(1+4i)", want: 1 + 4i},
+	{s: "(4i+1)", want: 1 + 4i},
+	{s: "(+1+4i)", want: 1 + 4i},
+	{s: "(+4i+1)", want: 1 + 4i},
+	{s: "(1e-4+1i)", want: 1e-4 + 1i},
+	{s: "(1e-4+i)", want: 1e-4 + 1i},
+	{s: "(1e-4-i)", want: 1e-4 - 1i},
+	{s: "(1e-4i-1)", want: -1 + 1e-4i},
+	{s: "(1e-4i+1)", want: 1 + 1e-4i},
+	{s: "NaN", want: cmplx.NaN()},
+	{s: "nan", want: cmplx.NaN()},
+	{s: "Inf", want: cmplx.Inf()},
+	{s: "inf", want: cmplx.Inf()},
+	{s: "(Inf+Infi)", want: complex(math.Inf(1), math.Inf(1))},
+	{s: "(-Inf+Infi)", want: complex(math.Inf(-1), math.Inf(1))},
+	{s: "(+Inf-Infi)", want: complex(math.Inf(1), math.Inf(-1))},
```

Comment:
what about `(-Inf-Infi)`, `(-inf-infi)` and `(nan-infi)` (+ other combinations)?
do we allow mixed case? _e.g._ `(+Inf-infi)`, `(NaN-nani)`?

Reply:
These tests aren't finished yet, Those cases don't mean much, but I will add them.

Reply:
On second thoughts I won't because of this section of the definition of [`cmplx.IsNaN`](https://golang.org/pkg/math/cmplx/#IsNaN) and how equality works with complex NaN values.